dmscripts/helpers/supplier_data_helpers.py

```diff
@@ -206,7 +206,7 @@ def remove_declaration_from_failed_applicants(self):
         """
```

Comment:
I can always remove them, or change them to list all possible types. I don't think we'll start using them unless we start, if you see what I mean. Otherwise it's a mission-long job to add types to every piece of code, and that's a job that'll never get done.

I think types help when you're writing code, because then you can be confident that you're passing what the author expects you to pass and the code becomes more self-documenting.

@katstevens , @risicle what do you think? Better typing (because I accept I've not included all possible types) or no typing?

Reply:
I think I broadly agree with the various sentiments here. I like typing, but we should be careful not to let ourselves get hung up on the whole thing. Certainly if you can use a general base type as your argument type this is good (don't enforce a `List` where all you need is a sequence, or maybe even an iterable will do). Compound types can get ... complex but I guess can be very illustrative.

When you get into the whole "should we accept a string where we could accept an int?" question gets a lot thornier around the various viewpoints of what "duck typing" really _is_ in the context of python, and whether you should force type conversion and the associated error handling at the border of your interface or whether you should be liberal.... it's all something which I'm on multiple fences about. For the particular example of `supplier_id` etc there are of course examples where we've got stung from id-ints-as-strings all too easily creeping into our data model and so maybe we should encourage conversion of these to ints ASAP?

I suspect we might just find ourselves having to get stung a bit before we discover what the best thing to do is from a _practical_ perspective.

Reply:
My onion: if we're going to try this out properly, I'd rather we start somewhere with more consistent code (and with better test coverage) than the scripts repo. It might be more useful in our versioned libraries, where we do want to specify a stricter contract about how we use functions, methods etc. But in theory I would be OK with introducing it in a planned way, if it's something we agree we want to do.

Reply:
Onion? Opinion! Friday...